File: src/Samples/CSharpApp/Views/Controls.cs
namespace CSharpApp.Views
{
    using System.Windows;
    using System.Windows.Controls;
    using System.Windows.Data;
    using System.Windows.Input;
    using CSharpApp.ViewModels;

    public class SubmitTextBox : TextBox
    {
        public SubmitTextBox()
            : base()
        {
            PreviewKeyDown += new KeyEventHandler(SubmitTextBox_PreviewKeyDown);
        }

        void SubmitTextBox_PreviewKeyDown(object sender, KeyEventArgs e)
        {
            if (e.Key == Key.Enter)
            {
                BindingExpression be = GetBindingExpression(TextBox.TextProperty);
                if (be != null)
                {
                    be.UpdateSource();
                }
            }
        }
    }

    public class VertexDataTemplateSelector : DataTemplateSelector
    {
        public VertexDataTemplateSelector() { }

        public override DataTemplate SelectTemplate(object item, DependencyObject container)
        {
            FrameworkElement element = container as FrameworkElement;
            if (element != null && item != null && item is INodeVertex)
            {
                INodeVertex vertexItem = (INodeVertex)item;
                if (vertexItem.Node.IsInput)
                {
                    return element.FindResource("InputVertexTemplate") as DataTemplate;
                }
                else
                {
                    return element.FindResource("OutputVertexTemplate") as DataTemplate;
                }
            }
            else
            {
                return null;
            }
        }
    }
}

File: src/Samples/FSharpApp/Views/Controls.fs
namespace FSharpApp.Views
open System.Windows
open System.Windows.Input
open System.Windows.Controls
open Utopia.Helpers
open FSharpApp.ViewModels

type VertexDataTemplateSelector() = 
    inherit DataTemplateSelector()
    override this.SelectTemplate(item : obj, container : DependencyObject) : DataTemplate = 
        let element = castAs<FrameworkElement>(container)
        if (element <> null && item <> null && (item |> isType<INodeVertex>)) then 
            let vertexItem = item :?> INodeVertex
            if (vertexItem.Node.IsInput) then element.FindResource("InputVertexTemplate") |> castAs<DataTemplate>
            else element.FindResource("OutputVertexTemplate") |> castAs<DataTemplate>
        else null

type SubmitTextBox() as this =
    inherit TextBox()
    do
        let previewKeyDown (sender : obj) (e : KeyEventArgs) =
                if (e.Key = Key.Enter) then
                    let be = this.GetBindingExpression(TextBox.TextProperty)
                    if be <> null then be.UpdateSource()
        this.PreviewKeyDown.AddHandler(fun sender e -> previewKeyDown sender e)